File: backend/models/Notification.js
const mongoose = require('mongoose');
const { sendPushNotification } = require('../services/firebaseService');

const notificationSchema = new mongoose.Schema({
    userId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: [true, 'L\'utilisateur est requis'],
        index: true,
    },
    type: {
        type: String,
        enum: [
            'reservationAssigned',
            'workerEnRoute',
            'workStarted',
            'workCompleted',
            'reservationCancelled',
            'paymentSuccess',
            'paymentFailed',
            'refundProcessed',
            'weatherAlert',
            'urgentRequest',
            'workerMessage',
            'systemNotification',
            'newMessage',      // Pour les messages de chat
            'tipReceived',     // Pourboire reçu par le déneigeur
            'rating',          // Évaluation reçue par le déneigeur
        ],
        required: [true, 'Le type est requis'],
    },
    title: {
        type: String,
        required: [true, 'Le titre est requis'],
        trim: true,
        maxlength: [100, 'Le titre ne peut pas dépasser 100 caractères'],
    },
    message: {
        type: String,
        required: [true, 'Le message est requis'],
        trim: true,
        maxlength: [500, 'Le message ne peut pas dépasser 500 caractères'],
    },
    priority: {
        type: String,
        enum: ['low', 'normal', 'high', 'urgent'],
        default: 'normal',
    },
    isRead: {
        type: Boolean,
        default: false,
        index: true,
    },
    reservationId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Reservation',
        default: null,
    },
    workerId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        default: null,
    },
    metadata: {
        type: mongoose.Schema.Types.Mixed,
        default: {},
    },
}, {
    timestamps: true,
});

// Index composé pour optimiser les requêtes
notificationSchema.index({ userId: 1, createdAt: -1 });
notificationSchema.index({ userId: 1, isRead: 1 });
notificationSchema.index({ userId: 1, isRead: 1, createdAt: -1 }); // Pour liste filtrée
notificationSchema.index({ createdAt: 1 }, { expireAfterSeconds: 90 * 24 * 60 * 60 }); // TTL 90 jours
notificationSchema.index({ type: 1, createdAt: -1 }); // Pour stats par type

// Méthode pour créer une notification
notificationSchema.statics.createNotification = async function(data) {
    const notification = new this(data);
    await notification.save();

    // Envoyer une push notification si l'utilisateur a un token FCM
    try {
        const User = mongoose.model('User');
        const user = await User.findById(data.userId).select('fcmToken notificationSettings email');

        // Logging détaillé pour le debugging
        const pushContext = {
            userId: data.userId?.toString(),
            notificationType: data.type,
            hasFcmToken: !!user?.fcmToken,
            pushEnabled: user?.notificationSettings?.pushEnabled !== false,
        };

        if (!user) {
            console.warn('[FCM] Utilisateur non trouvé pour notification push:', pushContext);
        } else if (!user.fcmToken) {
            // Log silencieux - l'utilisateur n'a simplement pas de token (normal si pas d'app installée)
            if (process.env.NODE_ENV !== 'production') {
                console.debug('[FCM] Utilisateur sans token FCM:', { ...pushContext, email: user.email });
            }
        } else if (user.notificationSettings?.pushEnabled === false) {
            console.debug('[FCM] Notifications push désactivées pour:', { ...pushContext, email: user.email });
        } else {
            const pushData = {
                notificationId: notification._id.toString(),
                type: data.type,
                ...(data.reservationId && { reservationId: data.reservationId.toString() }),
                ...(data.workerId && { workerId: data.workerId.toString() }),
            };

            const result = await sendPushNotification(
                user.fcmToken,
                data.title,
                data.message,
                pushData
            );

            // Si le token est invalide, le supprimer
            if (result && result.invalidToken) {
                await User.findByIdAndUpdate(data.userId, { fcmToken: null });
                console.warn('[FCM] Token invalide supprimé pour utilisateur:', {
                    userId: data.userId?.toString(),
                    email: user.email,
                    errorCode: result.errorCode,
                });
            } else if (result && result.success) {
                console.debug('[FCM] Push envoyée avec succès:', pushContext);
            }
        }
    } catch (error) {
        console.error('[FCM] Erreur envoi notification push:', {
            userId: data.userId?.toString(),
            type: data.type,
            error: error.message,
            stack: process.env.NODE_ENV !== 'production' ? error.stack : undefined,
        });
        // Ne pas faire échouer la création de notification si la push échoue
    }

    return notification;
};

// Méthode helper pour créer des notifications spécifiques
notificationSchema.statics.notifyReservationAssigned = async function(reservation, worker) {
    return await this.createNotification({
        userId: reservation.userId,
        type: 'reservationAssigned',
        title: 'Déneigeur assigné',
        message: `${worker.firstName} ${worker.lastName} a accepté votre demande de déneigement`,
        priority: 'high',
        reservationId: reservation._id,
        workerId: worker._id,
        metadata: {
            workerName: `${worker.firstName} ${worker.lastName}`,
            departureTime: reservation.departureTime,
        },
    });
};

notificationSchema.statics.notifyWorkerEnRoute = async function(reservation, worker) {
    return await this.createNotification({
        userId: reservation.userId,
        type: 'workerEnRoute',
        title: 'Déneigeur en route',
        message: `${worker.firstName} est en route vers votre véhicule`,
        priority: 'high',
        reservationId: reservation._id,
        workerId: worker._id,
        metadata: {
            workerName: `${worker.firstName} ${worker.lastName}`,
            estimatedArrival: new Date(Date.now() + 15 * 60000), // +15 min
        },
    });
};

notificationSchema.statics.notifyWorkStarted = async function(reservation, worker) {
    return await this.createNotification({
        userId: reservation.userId,
        type: 'workStarted',
        title: 'Déneigement commencé',
        message: `${worker.firstName} a commencé le déneigement de votre véhicule`,
        priority: 'normal',
        reservationId: reservation._id,
        workerId: worker._id,
        metadata: {
            startTime: new Date(),
        },
    });
};

notificationSchema.statics.notifyWorkCompleted = async function(reservation, worker) {
    return await this.createNotification({
        userId: reservation.userId,
        type: 'workCompleted',
        title: 'Déneigement terminé',
        message: `${worker.firstName} a terminé le déneigement. Votre véhicule est prêt!`,
        priority: 'high',
        reservationId: reservation._id,
        workerId: worker._id,
        metadata: {
            completedAt: reservation.completedAt,
            totalPrice: reservation.totalPrice,
        },
    });
};

notificationSchema.statics.notifyPaymentSuccess = async function(reservation) {
    return await this.createNotification({
        userId: reservation.userId,
        type: 'paymentSuccess',
        title: 'Paiement réussi',
        message: `Votre paiement de ${reservation.totalPrice.toFixed(2)} $ a été effectué avec succès`,
        priority: 'normal',
        reservationId: reservation._id,
        metadata: {
            amount: reservation.totalPrice,
            paymentIntentId: reservation.paymentIntentId,
        },
    });
};

notificationSchema.statics.notifyPaymentFailed = async function(reservation, error) {
    return await this.createNotification({
        userId: reservation.userId,
        type: 'paymentFailed',
        title: 'Paiement échoué',
        message: `Le paiement de ${reservation.totalPrice.toFixed(2)} $ a échoué. Veuillez réessayer.`,
        priority: 'urgent',
        reservationId: reservation._id,
        metadata: {
            amount: reservation.totalPrice,
            error: error,
        },
    });
};

notificationSchema.statics.notifyReservationCancelled = async function(reservation, reason) {
    return await this.createNotification({
        userId: reservation.userId,
        type: 'reservationCancelled',
        title: 'Réservation annulée',
        message: reason || 'Votre réservation a été annulée',
        priority: 'high',
        reservationId: reservation._id,
        metadata: {
            cancelledAt: reservation.cancelledAt,
            reason: reason,
        },
    });
};

module.exports = mongoose.model('Notification', notificationSchema);

File: lib/features/notifications/domain/entities/notification.dart
import 'package:equatable/equatable.dart';

import '../../../../l10n/app_localizations.dart';

enum NotificationType {
  reservationAssigned, // Déneigeur a accepté la tâche
  workerEnRoute, // Déneigeur est en route
  workStarted, // Travail a commencé
  workCompleted, // Travail terminé
  reservationCancelled, // Réservation annulée
  paymentSuccess, // Paiement réussi
  paymentFailed, // Paiement échoué
  refundProcessed, // Remboursement effectué
  weatherAlert, // Alerte météo (neige prévue)
  urgentRequest, // Demande urgente
  workerMessage, // Message du déneigeur
  newMessage, // Nouveau message de chat
  tipReceived, // Pourboire reçu par le déneigeur
  rating, // Évaluation reçue par le déneigeur
  systemNotification, // Notification système
}

enum NotificationPriority {
  low,
  normal,
  high,
  urgent,
}

class AppNotification extends Equatable {
  final String id;
  final String userId;
  final NotificationType type;
  final String title;
  final String message;
  final NotificationPriority priority;
  final bool isRead;
  final DateTime createdAt;
  final String? reservationId;
  final String? workerId;
  final Map<String, dynamic>? metadata;

  const AppNotification({
    required this.id,
    required this.userId,
    required this.type,
    required this.title,
    required this.message,
    this.priority = NotificationPriority.normal,
    this.isRead = false,
    required this.createdAt,
    this.reservationId,
    this.workerId,
    this.metadata,
  });

  // Business logic
  bool get isUrgent => priority == NotificationPriority.urgent;

  bool get isRecent {
    final now = DateTime.now();
    final difference = now.difference(createdAt);
    return difference.inHours < 24;
  }

  bool get isToday {
    final now = DateTime.now();
    return createdAt.year == now.year &&
        createdAt.month == now.month &&
        createdAt.day == now.day;
  }

  String get timeAgo {
    final now = DateTime.now();
    final difference = now.difference(createdAt);

    if (difference.inMinutes < 1) {
      return 'À l\'instant';
    } else if (difference.inMinutes < 60) {
      return 'Il y a ${difference.inMinutes} min';
    } else if (difference.inHours < 24) {
      return 'Il y a ${difference.inHours} h';
    } else if (difference.inDays < 7) {
      return 'Il y a ${difference.inDays} j';
    } else {
      return 'Il y a ${difference.inDays ~/ 7} sem';
    }
  }

  String localizedTimeAgo(AppLocalizations l10n) {
    final now = DateTime.now();
    final difference = now.difference(createdAt);
    if (difference.inMinutes < 1) {
      return l10n.notif_justNow;
    } else if (difference.inMinutes < 60) {
      return l10n.notif_minutesAgo(difference.inMinutes);
    } else if (difference.inHours < 24) {
      return l10n.notif_hoursAgo(difference.inHours);
    } else if (difference.inDays < 7) {
      return l10n.notif_daysAgo(difference.inDays);
    } else {
      return l10n.notif_weeksAgo(difference.inDays ~/ 7);
    }
  }

  AppNotification copyWith({
    String? id,
    String? userId,
    NotificationType? type,
    String? title,
    String? message,
    NotificationPriority? priority,
    bool? isRead,
    DateTime? createdAt,
    String? reservationId,
    String? workerId,
    Map<String, dynamic>? metadata,
  }) {
    return AppNotification(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      type: type ?? this.type,
      title: title ?? this.title,
      message: message ?? this.message,
      priority: priority ?? this.priority,
      isRead: isRead ?? this.isRead,
      createdAt: createdAt ?? this.createdAt,
      reservationId: reservationId ?? this.reservationId,
      workerId: workerId ?? this.workerId,
      metadata: metadata ?? this.metadata,
    );
  }

  AppNotification markAsRead() {
    return copyWith(isRead: true);
  }

  @override
  List<Object?> get props => [
        id,
        userId,
        type,
        title,
        message,
        priority,
        isRead,
        createdAt,
        reservationId,
        workerId,
        metadata,
      ];
}

extension NotificationTypeExtension on NotificationType {
  String get iconName {
    switch (this) {
      case NotificationType.reservationAssigned:
        return 'person_add';
      case NotificationType.workerEnRoute:
        return 'directions_car';
      case NotificationType.workStarted:
        return 'construction';
      case NotificationType.workCompleted:
        return 'check_circle';
      case NotificationType.reservationCancelled:
        return 'cancel';
      case NotificationType.paymentSuccess:
        return 'payment';
      case NotificationType.paymentFailed:
        return 'error';
      case NotificationType.refundProcessed:
        return 'money_off';
      case NotificationType.weatherAlert:
        return 'wb_cloudy';
      case NotificationType.urgentRequest:
        return 'priority_high';
      case NotificationType.workerMessage:
        return 'message';
      case NotificationType.newMessage:
        return 'chat_bubble';
      case NotificationType.tipReceived:
        return 'attach_money';
      case NotificationType.rating:
        return 'star';
      case NotificationType.systemNotification:
        return 'info';
    }
  }

  String localizedDisplayName(AppLocalizations l10n) {
    switch (this) {
      case NotificationType.reservationAssigned:
        return l10n.notifType_reservationAssigned;
      case NotificationType.workerEnRoute:
        return l10n.notifType_workerEnRoute;
      case NotificationType.workStarted:
        return l10n.notifType_workStarted;
      case NotificationType.workCompleted:
        return l10n.notifType_workCompleted;
      case NotificationType.reservationCancelled:
        return l10n.notifType_reservationCancelled;
      case NotificationType.paymentSuccess:
        return l10n.notifType_paymentSuccess;
      case NotificationType.paymentFailed:
        return l10n.notifType_paymentFailed;
      case NotificationType.refundProcessed:
        return l10n.notifType_refundProcessed;
      case NotificationType.weatherAlert:
        return l10n.notifType_weatherAlert;
      case NotificationType.urgentRequest:
        return l10n.notifType_urgentRequest;
      case NotificationType.workerMessage:
        return l10n.notifType_workerMessage;
      case NotificationType.newMessage:
        return l10n.notifType_newMessage;
      case NotificationType.tipReceived:
        return l10n.notifType_tipReceived;
      case NotificationType.rating:
        return l10n.notifType_rating;
      case NotificationType.systemNotification:
        return l10n.notifType_systemNotification;
    }
  }
}
